src/auth.rs

```diff
@@ -0,0 +1,155 @@
+        // There are a couple of layers of signing involved here, that eventually
+        // chain back to a secret key shared between this storage node and the tokenserver.
+        // The Authorization header will look like this:
+        //    Authorization: Hawk id="<...>", ts="1353832234", nonce="j4h3g2", mac="6R4rV5iE+NPoym+WwjeHzjAGXUtLNIxmo1vpMofpLAE="
+        // Where the "id" field is a signed-and-base64-encoded JSON blob of user metadata,
+        // produced by the python "tokenlib" library (https://github.com/mozilla-services/tokenlib)
+        // The decoding procedure must proceed as follows:
+
+        let id = header.id.ok_or(AuthError)?;
+        let mac = header.mac.ok_or(AuthError)?;
+
+        //   * Obtain the `master_secret` master token secret from config
+        //   * Derive `signing_secret = HKDF-SHA256(master_secret, size=32, salt=None, info="services.mozilla.com/tokenlib/v1/signing")`
+        let hkdf = Hkdf::<Sha256>::extract(None, &hex::decode(&settings.master_secret)?);
+        let mut signing_secret = [ 0u8; 32 ];
+        hkdf.expand(b"services.mozilla.com/tokenlib/v1/signing", &mut signing_secret)?;
+
+        //   * Extract the `id` from the Hawk auth header
+        //   * urlsafe_b64decode `id` and split off the last 32 bytes to give (`payload`, `signature`)
+        let decoded_id = base64::decode(&id)?;
+        let payload_length = decoded_id.len() - 32;
+        let payload = &decoded_id[0..payload_length];
+        let signature = &decoded_id[payload_length..];
+
+        //   * Calculate `HMAC-SHA256(payload, signing_secret)` and check that it matches `signature`
+        let mut check: Hmac<Sha256> = Hmac::new_varkey(&signing_secret)?;
+        check.input(&payload);
+        let expected_signature = check.result().code();
+        if !constant_time_compare(signature, &expected_signature) {
+            return Err(AuthError);
+        }
+
+        //   * JSON decode `payload` to give an object like: {
+        //       'userid': 42,
+        //       'expires': 1329875384.073159
+        //       'salt': '1c033f'
+        //     }
+        let payload: HawkPayload = serde_json::from_slice(payload)?;
+
+        //   * Check that the "expires" timestamp is not in the past.
+        let now = Utc::now().timestamp_millis();
+        if (payload.expires.round() as i64) < now {
+            return Err(AuthError);
+        }
+
+        //   * Derive `token_secret = HKDF-SHA256(master_secret, size=32, salt=payload["salt"], info="services.mozilla.com/tokenlib/v1/derive/" + id)`
+        let hkdf: Hkdf<Sha256> = Hkdf::extract(Some(payload.salt.as_bytes()), &hex::decode(&settings.master_secret)?);
+        let mut token_secret = [ 0u8; 32 ];
+        hkdf.expand(format!("services.mozilla.com/tokenlib/v1/derive/{}", id).as_bytes(), &mut token_secret)?;
+
+        //   * Use `token_secret` as the secret key for calculating the Hawk request MAC
```

Comment:
I wasn't sure about the significance of `calculating the Hawk request MAC` here. Does it mean something more than the direct comparison of `mac` and `token_secret` I'm doing below?